registrations/api.py

```diff
@@ -220,14 +221,165 @@ def send_invitation(self, request, pk=None, version=None):
+                    email=user.email
+                ).exists()
+                or registration.publisher in registration_admin_organizations
+            ):
+                raise DRFPermissionDenied(
+                    _(
+                        "Only the admins of the organization {organization} have access rights."
+                    ).format(organization=registration.publisher)
+                )
+
+            registrations.append(registration)
+
+        return queryset.filter(**{f"{name}__in": registrations})
+
+    @staticmethod
+    def _build_text_annotations(relation_accessor=""):
+        return {
+            f"{relation_accessor}first_last_name": Concat(
+                f"{relation_accessor}first_name",
+                Value(" "),
+                f"{relation_accessor}last_name",
+            ),
+            f"{relation_accessor}last_first_name": Concat(
+                f"{relation_accessor}last_name",
+                Value(" "),
+                f"{relation_accessor}first_name",
+            ),
+        }
+
+    @staticmethod
+    def _build_text_filter(text_param, relation_accessor=""):
+        filters = {
+            f"{relation_accessor}first_last_name__icontains": text_param,
+            f"{relation_accessor}last_first_name__icontains": text_param,
+            f"{relation_accessor}email__icontains": text_param,
+            f"{relation_accessor}extra_info__icontains": text_param,
+            f"{relation_accessor}membership_number__icontains": text_param,
+            f"{relation_accessor}phone_number__icontains": text_param,
+        }
+
+        q_set = Q()
+        for item in filters.items():
+            q_set |= Q(item)
+
+        return q_set
+
+    def filter_text(self, queryset, name, value):
+        return queryset.annotate(**self._build_text_annotations()).filter(
+            self._build_text_filter(value)
+        )
+
+
+class SignUpFilter(SignUpBaseFilter):
+    class Meta:
+        model = SignUp
+        fields = ("registration", "attendee_status")
+
+
+class SignUpGroupFilter(SignUpBaseFilter):
+    attendee_status = django_filters.MultipleChoiceFilter(
+        field_name="signups__attendee_status",
+        choices=SignUp.ATTENDEE_STATUSES,
+        widget=django_filters.widgets.CSVWidget(),
+    )
+
+    def filter_text(self, queryset, name, value):
+        return queryset.annotate(
+            **self._build_text_annotations(relation_accessor="signups__")
+        ).filter(self._build_text_filter(value, relation_accessor="signups__"))
+
+    class Meta:
+        model = SignUpGroup
+        fields = ("registration",)
+
+
 class SignUpViewSet(
     UserDataSourceAndOrganizationMixin,
     viewsets.ModelViewSet,
 ):
     serializer_class = SignUpSerializer
     queryset = SignUp.objects.all()
+    filter_backends = [SignUpFilterBackend]
+    filterset_class = SignUpFilter
     permission_classes = [CanAccessSignup & DataSourceResourceEditPermission]
 
+    def get_filterset_kwargs(self):
```

Comment:
Why we need this? I'm probably missing something 🤔

Reply:
@nicobav I predicted that this would be questioned :-) This is because we are passing `action` from the View as a kwarg to the Filterset and the reason for that is that I wanted to keep business logic untouched after this refactoring. The existing code gave a 403 error if any other action that `list` was done and if the user wasn't allowed to do the action on the objects. It did that by checking the `action` firstly so I kept it as it was (as otherwise the endpoint would give a 404 as the object would not be included in the queryset). The 403 is perhaps semantically better to the user than 404 in this case.

Reply:
I think you can also do this purely in `SignUpFilterBackend.get_filterset_kwargs` since it gets `view` as an argument and that is the instance of the view, so in `SignUpFilterBackend` you could just have:

```
    def get_filterset_kwargs(self, request, queryset, view):
        kwargs = super().get_filterset_kwargs(request, queryset, view)
        kwargs["action"] = view.action

        return kwargs
```

Reply:
Good point -> now changed to work like that